src/main/java/org/opensearch/neuralsearch/highlight/SemanticHighlightingResponseProcessor.java

```diff
@@ -0,0 +1,390 @@
+            responseListener.onResponse(response);
+        } else {
+            responseListener.onFailure(error);
+        }
+    }
+
+    /**
+     * Process requests sequentially - one at a time
+     */
+    private void processSequentialRequests(
+        List<SentenceHighlightingRequest> requests,
+        List<SearchHit> validHits,
+        int currentIndex,
+        SearchResponse response,
+        String semanticHighlightField,
+        ActionListener<SearchResponse> responseListener,
+        long startTime
+    ) {
+        // Base case: all requests processed
+        if (currentIndex >= requests.size()) {
+            long totalTime = System.currentTimeMillis() - startTime;
+            SearchResponse finalResponse = ProcessorUtils.updateResponseTookTime(response, totalTime);
+            responseListener.onResponse(finalResponse);
+            return;
+        }
+
+        // Process current request
+        mlClientAccessor.inferenceSentenceHighlighting(requests.get(currentIndex), ActionListener.wrap(highlightResults -> {
+            try {
+                applySingleHighlightResults(validHits.get(currentIndex), highlightResults, semanticHighlightField);
+                // Process next request recursively
+                processSequentialRequests(
+                    requests,
+                    validHits,
+                    currentIndex + 1,
+                    response,
+                    semanticHighlightField,
+                    responseListener,
+                    startTime
+                );
+            } catch (Exception e) {
+                handleProcessingError(e, response, responseListener);
+            }
+        }, error -> {
+            log.error(String.format(Locale.ROOT, "sequential inference failed for request %d", currentIndex), error);
+            handleProcessingError(error, response, responseListener);
+        }));
+    }
+
+    /**
+     * Apply highlights to a specific search hit
+     */
+    private void applyHighlightsToHit(SearchHit hit, List<Map<String, Object>> highlights, String fieldName) {
+        try {
+            Map<String, Object> source = hit.getSourceAsMap();
+            if (source == null) {
+                return;
+            }
+
+            String text = (String) source.get(fieldName);
+            if (text != null) {
+                String highlightedText = applyPositionHighlights(text, highlights);
+
+                Map<String, HighlightField> highlightFields = Optional.ofNullable(hit.getHighlightFields()).orElse(new HashMap<>());
+
+                HighlightField highlightField = new HighlightField(fieldName, new Text[] { new Text(highlightedText) });
+                highlightFields.put(fieldName, highlightField);
+                hit.highlightFields(highlightFields);
+            }
+        } catch (Exception e) {
+            log.error(String.format(Locale.ROOT, "error applying highlights to hit %s", hit.getId()), e);
```

Comment:
You don't need to use String.format for logging.
```suggestion
            log.error("error applying highlights to hit {}", hit.getId()), e);
```